platforms/core-configuration/declarative-dsl-provider/src/main/kotlin/org/gradle/internal/declarativedsl/project/dependencyConfigurationSchema.kt

```diff
@@ -103,6 +133,48 @@ class RuntimeDependencyFunctionResolver(configurations: DependencyConfigurations
                 }
             })
         }
+
         return RuntimeFunctionResolver.Resolution.Unresolved
     }
 }
+
+
+private
+class ImplicitDependencyCollectorFunctionResolver(configurations: DependencyConfigurations) : RuntimeFunctionResolver {
```

Comment:
🤔 I would suggest making the runtime resolver more strict by passing it the information collected at schema building time and using that to make decisions resolving a function – rather than inspecting the receiver type at runtime again.

That information might include the types that have their dependency collectors detected, as well as the "configuration" (dependency collector property) names. 

The downside of doing this is would be the more complex wiring between the two parts, like using another object to "register" detected dependency collectors when `ImplicitDependencyCollectorFunctionExtractor` finds them, and then use that object as the source of information for `ImplicitDependencyCollectorFunctionResolver`. However, I think that making this connection explicit rather than duplicating the discovery would improve maintainability of the implementation.

WDYT?

Reply:
As I see it, in the future, such collected data might come along with the schema as custom extras, for the purposes of caching and reuse in the DSL runtime when the schema is unchanged.

Reply:
Sounds like a good idea.  My focus is going to move back to the `declarative-prototypes` side next.  My hope was that this is good enough to merge so that we could dep master from that project instead of the snapshot I'm trying to publish.

What would you say to merging this as-is (or with any minor improvements you think are critical) and circling back later to do this and process properties?

Reply:
I'm fine with merging this as-is and making improvements later asynchronously. I can help with some of the proposed changes, too.